File: tagging_tools/SOCTagger.py
import yaml
import TaggingUtils

def generate_soc_ranges(soc_file, range_file, policy_inits):
    soc_cfg = yaml.load(soc_file.read(), Loader=yaml.FullLoader)

    soc_ranges = {}
    for elem in soc_cfg['SOC']:
        elem_fields = soc_cfg['SOC'][elem]
        name = elem_fields['name']
        start = elem_fields['start']
        end = elem_fields['end']
        soc_ranges[name] = (start, end)

    # filter SOC range entries by those required by the policy init
    policy_soc = policy_inits['Require']['SOC']
    for device_type in policy_soc:
        for elem in policy_soc[device_type]:
            name = "SOC.{}.{}".format(device_type, elem)
            if name in soc_ranges:
                print('{0}: 0x{1:X} - 0x{2:X}'.format(name, soc_ranges[name][0], soc_ranges[name][1]))
                range_file.write_range(soc_ranges[name][0], soc_ranges[name][1], name)

File: tagging_tools/soc_tagger.cc
#include <cstdint>
#include <iostream>
#include <map>
#include <string>
#include <utility>
#include <yaml-cpp/yaml.h>
#include "metadata_memory_map.h"
#include "range_map.h"
#include "reporter.h"

namespace policy_engine {

void add_soc_ranges(range_map_t& range_map, const std::string& soc_file, const YAML::Node& policy_inits, reporter_t& err) {
  YAML::Node soc_cfg = YAML::LoadFile(soc_file);

  std::map<std::string, range_t> soc_ranges;
  for (const auto& elem : soc_cfg["SOC"])
    soc_ranges[elem.second["name"].as<std::string>()] = range_t{elem.second["start"].as<uint64_t>(), elem.second["end"].as<uint64_t>()};
  for (const auto& device : policy_inits["Require"]["SOC"]) {
    for (const auto& elem : device.second) {
      std::string name = "SOC." + device.first.as<std::string>() + "." + elem.first.as<std::string>();
      if (soc_ranges.find(name) != soc_ranges.end()) {
        err.info("%s: %#lx - %#lx\n", name, soc_ranges[name].start, soc_ranges[name].end);
        range_map.add_range(soc_ranges[name].start, soc_ranges[name].end, name);
      }
    }
  }
}
    
} // namespace policy_engine